core/src/main/scala/akka/kafka/internal/TransactionalSource.scala

```diff
@@ -36,9 +82,85 @@ private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerS
     IsolationLevel.READ_COMMITTED.toString.toLowerCase(Locale.ENGLISH)
   )
 
+  case object Drained
+  case class Drain[T](drainedConfirmationRef: Option[ActorRef], drainedConfirmationMsg: T)
+  case class Committed(offsets: Map[TopicPartition, OffsetAndMetadata])
+
   override protected def logic(shape: SourceShape[TransactionalMessage[K, V]]): GraphStageLogic with Control =
     new SingleSourceLogic[K, V, TransactionalMessage[K, V]](shape, txConsumerSettings, subscription)
     with TransactionalMessageBuilder[K, V] {
+      var inFlightRecords = InFlightRecords.empty
+
+      override def messageHandling = super.messageHandling.orElse(drainHandling).orElse {
+        case (_, Revoked(tps)) =>
+          inFlightRecords.revoke(tps.toSet)
+      }
+
+      override def shuttingDownReceive = super.shuttingDownReceive.orElse(drainHandling)
+
+      def drainHandling: PartialFunction[(ActorRef, Any), Unit] = {
+        case (sender, Committed(offsets)) =>
+          inFlightRecords.committed(offsets.mapValues(_.offset() - 1))
+          sender ! Done
+        case (sender, Drain(ack, msg)) =>
+          if (inFlightRecords.empty()) {
+            log.debug("Source drained")
+            ack.getOrElse(sender) ! msg
+          } else {
+            log.debug(s"Draining partitions {}", inFlightRecords)
+            // TODO: This timeout should be somehow related to the committing interval.
+            // For instance having 10.millis makes no sense if we commit transaction every second
+            materializer.scheduleOnce(30.millis, new Runnable {
+              override def run(): Unit =
+                sourceActor.ref ! Drain(ack.orElse(Some(sender)), msg)
+            })
+          }
+      }
+
       override def groupId: String = txConsumerSettings.properties(ConsumerConfig.GROUP_ID_CONFIG)
+      lazy val committedMarker: CommittedMarker = {
+        val ec = materializer.executionContext
+        CommittedMarkerRef(sourceActor.ref, commitTimeout)(ec)
+      }
+
+      override def onMessage(rec: ConsumerRecord[K, V]): Unit =
+        inFlightRecords.add(Map(new TopicPartition(rec.topic(), rec.partition()) -> rec.offset()))
+
+      override protected def stopConsumerActor(): Unit =
+        sourceActor.ref.tell(Drain(Some(consumerActor), KafkaConsumerActor.Internal.Stop), sourceActor.ref)
+
+      // This is invoked in the KafkaConsumerActor thread when doing poll.
+      override def partitionRevokedCB(revokedTps: Set[TopicPartition]): Unit = {
+        if (waitForDraining()) {
+          sourceActor.ref ! Revoked(revokedTps.toList)
+        } else {
+          sourceActor.ref ! Failure(new Error("Timeout while draining"))
+        }
+        super.partitionRevokedCB(revokedTps)
+      }
+
+      def waitForDraining(): Boolean = {
+        import akka.pattern.ask
+        // Shall we use commitTimeout here?
```

Comment:
Yes, it should be at least commit timeout. Possible a bit larger as well. Can you refactor this from hardcoded constant to configuration value?

Reply:
After some thoughts I think using commitTimeout is just ok, will use that.